### ISSUE TITLE
generic buffer in DAL

### ISSUE THREAD

**Issue by mmoskal** [lancaster-university/microbit-dal#112]
Currently, we have ManagedString (immutable) and PacketBuffer (with additional RSSI field). I think it would make sense to have a generic ManagedBuffer, and possibly drop PacketBuffer (and pass RSSI some other way; most applications won't care anyway). I imagine it would be useful in other APIs as well.

I have an example implementation here: https://github.com/Microsoft/kindscript-microbit-core/blob/master/inc/ManagedBuffer.h

It comes with additional utility methods for filling, copying, shifting and rotation. These are all useful for example in a neopixel driver.

We then extend it with functions to read/write numbers in different sizes/endianess. Note sure if this is appropriate in DAL:

https://github.com/Microsoft/kindscript-microbit/blob/master/libs/microbit/buffer.cpp

BTW, the payload field in PacketBuffer is not aligned. I imagine this can lead to trouble. If we keep it, maybe we can change RSSI to one byte.

CC @pelikhan @finneyj @jamesadevine

**Comment by finneyj** [comment]
Thanks @mmoskal - great suggestion. We certainly don't want to have a whole runaway of refcounted types without any consistency. I think a generic buffer would definitely be a "good thing" (tm).

Would you mind if we look at upstreaming your ManagedBuffer implementation?Looks like a really good basis!

**Comment by mmoskal** [comment]
Of course, go ahead! Not sure if you want the methods to read/write different sized ints, but maybe they just don't make much sense in C++ - you can always memcpy(). I was basing mine on this https://nodejs.org/api/buffer.html

**Comment by whs-che** [comment]
Hello - I can confirm this point:

"BTW, the payload field in PacketBuffer is not aligned. I imagine this can lead to trouble. If we keep it, maybe we can change RSSI to one byte."

It did cause me a problem when I mapped a structure directly over the returned user buffer. I fixed it initially by simply copying the data into a correctly aligned local version. A bit inefficient but it works!
I have since tried the suggested RSSI to one byte and that also works.

My final fix for this was to add a padding byte like so:
struct PacketData : RefCounted
{
    uint16_t        rssi;               // The radio signal strength this packet was received.
    uint8_t         length;             // The length of the payload in bytes
    uint8_t         pad1;
    uint8_t         payload[0];         // User / higher layer protocol data
};

My only logic for doing this is that the user payload will end up on the same boundary as if it was directly allocated using the DAL allocation routines. As a user that seems to be a reasonable but maybe naive assumption to make.  I'm trying to ignore the fact that the native routines allocate on 8 byte boundaries and I've just wasted a precious byte of memory!

Hope this is of use.
